### ISSUE TITLE
Network not learning

### ISSUE THREAD

**Issue by ValterFallenius** [openclimatefix/metnet#22]
**Training loss is not decreasing**
I have implemented the network in the PR "lightning" branch with pytorch lightning and tried to find any bugs. The network compiles without issues and seems to generate gradients but still network fails to learn anything. I have tried to play around with the learning rate and plot the data at different stages but even with 4 training samples (it should be able to overfit these) it fails to decrease the loss even after 100 epochs... 

Here is the training loss plotted:
![W B Chart 2022-03-05 11_20_29](https://user-images.githubusercontent.com/21970939/156879117-6978224f-6e63-47cd-b13b-f7bf02094963.png)

It seems like it's doing something but not nearly quick enough to overfit the small dataset. Something is wrong...

Hyperparameters:
n_samples = 4
hidden_dim=8, 
forecast_steps=1, 
input_channels=15, 
output_channels=6, #512
input_size=112, # 112
n_samples = 100,
num_workers = 8,
batch_size = 1,
learning_rate = 1e-2

Below is a weights&biases grad report. As you can see most gradients are non-zero, I'm not sure why image_encoder has very small gradients for their biases...

`wandb:                                                                    epoch 83
wandb:                                            grad_2.0_norm/head.bias_epoch 0.0746
wandb:                                             grad_2.0_norm/head.bias_step 0.049
wandb:                                          grad_2.0_norm/head.weight_epoch 0.0862
wandb:                                           grad_2.0_norm/head.weight_step 0.081
wandb:                   grad_2.0_norm/image_encoder.module.module.0.bias_epoch 0.0
wandb:                    grad_2.0_norm/image_encoder.module.module.0.bias_step 0.0
wandb:                 grad_2.0_norm/image_encoder.module.module.0.weight_epoch 0.06653
wandb:                  grad_2.0_norm/image_encoder.module.module.0.weight_step 0.043
wandb:                   grad_2.0_norm/image_encoder.module.module.2.bias_epoch 0.00017
wandb:                    grad_2.0_norm/image_encoder.module.module.2.bias_step 0.0001
wandb:                 grad_2.0_norm/image_encoder.module.module.2.weight_epoch 0.003
wandb:                  grad_2.0_norm/image_encoder.module.module.2.weight_step 0.0019
wandb:                   grad_2.0_norm/image_encoder.module.module.3.bias_epoch 0.0
wandb:                    grad_2.0_norm/image_encoder.module.module.3.bias_step 0.0
wandb:                 grad_2.0_norm/image_encoder.module.module.3.weight_epoch 0.16387
wandb:                  grad_2.0_norm/image_encoder.module.module.3.weight_step 0.1125
wandb:                   grad_2.0_norm/image_encoder.module.module.4.bias_epoch 0.00013
wandb:                    grad_2.0_norm/image_encoder.module.module.4.bias_step 0.0001
wandb:                 grad_2.0_norm/image_encoder.module.module.4.weight_epoch 0.00203
wandb:                  grad_2.0_norm/image_encoder.module.module.4.weight_step 0.0012
wandb:                   grad_2.0_norm/image_encoder.module.module.5.bias_epoch 0.0
wandb:                    grad_2.0_norm/image_encoder.module.module.5.bias_step 0.0
wandb:                 grad_2.0_norm/image_encoder.module.module.5.weight_epoch 0.15237
wandb:                  grad_2.0_norm/image_encoder.module.module.5.weight_step 0.1151
wandb:                   grad_2.0_norm/image_encoder.module.module.6.bias_epoch 0.0032
wandb:                    grad_2.0_norm/image_encoder.module.module.6.bias_step 0.0018
wandb:                 grad_2.0_norm/image_encoder.module.module.6.weight_epoch 0.00157
wandb:                  grad_2.0_norm/image_encoder.module.module.6.weight_step 0.0012
wandb:                   grad_2.0_norm/image_encoder.module.module.7.bias_epoch 0.00497
wandb:                    grad_2.0_norm/image_encoder.module.module.7.bias_step 0.003
wandb:                 grad_2.0_norm/image_encoder.module.module.7.weight_epoch 0.11753
wandb:                  grad_2.0_norm/image_encoder.module.module.7.weight_step 0.0915
wandb:    grad_2.0_norm/temporal_agg.0.axial_attentions.0.fn.to_kv.weight_epoch 0.03763
wandb:     grad_2.0_norm/temporal_agg.0.axial_attentions.0.fn.to_kv.weight_step 0.0277
wandb:     grad_2.0_norm/temporal_agg.0.axial_attentions.0.fn.to_out.bias_epoch 0.0412
wandb:      grad_2.0_norm/temporal_agg.0.axial_attentions.0.fn.to_out.bias_step 0.0289
wandb:   grad_2.0_norm/temporal_agg.0.axial_attentions.0.fn.to_out.weight_epoch 0.05167
wandb:    grad_2.0_norm/temporal_agg.0.axial_attentions.0.fn.to_out.weight_step 0.0369
wandb:     grad_2.0_norm/temporal_agg.0.axial_attentions.0.fn.to_q.weight_epoch 0.0008
wandb:      grad_2.0_norm/temporal_agg.0.axial_attentions.0.fn.to_q.weight_step 0.0008
wandb:    grad_2.0_norm/temporal_agg.0.axial_attentions.1.fn.to_kv.weight_epoch 0.04393
wandb:     grad_2.0_norm/temporal_agg.0.axial_attentions.1.fn.to_kv.weight_step 0.0216
wandb:     grad_2.0_norm/temporal_agg.0.axial_attentions.1.fn.to_out.bias_epoch 0.0412
wandb:      grad_2.0_norm/temporal_agg.0.axial_attentions.1.fn.to_out.bias_step 0.0289
wandb:   grad_2.0_norm/temporal_agg.0.axial_attentions.1.fn.to_out.weight_epoch 0.04287
wandb:    grad_2.0_norm/temporal_agg.0.axial_attentions.1.fn.to_out.weight_step 0.027
wandb:     grad_2.0_norm/temporal_agg.0.axial_attentions.1.fn.to_q.weight_epoch 0.0014
wandb:      grad_2.0_norm/temporal_agg.0.axial_attentions.1.fn.to_q.weight_step 0.0009
wandb:            grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_h1.bias_epoch 0.00197
wandb:             grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_h1.bias_step 0.001
wandb:          grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_h1.weight_epoch 0.03313
wandb:           grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_h1.weight_step 0.0216
wandb:            grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_h2.bias_epoch 0.00103
wandb:             grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_h2.bias_step 0.0004
wandb:          grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_h2.weight_epoch 0.00353
wandb:           grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_h2.weight_step 0.002
wandb:            grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_zr.bias_epoch 0.00133
wandb:             grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_zr.bias_step 0.0009
wandb:          grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_zr.weight_epoch 0.02123
wandb:           grad_2.0_norm/temporal_enc.rnn.cell_list.0.conv_zr.weight_step 0.0147
wandb:                                                grad_2.0_norm_total_epoch 0.31513
wandb:                                                 grad_2.0_norm_total_step 0.2254
wandb:                                                         train/loss_epoch 1.72826
wandb:                                                          train/loss_step 1.73303
wandb:                                                      trainer/global_step 251
wandb:                                                    validation/loss_epoch 1.76064
`

I have plotted the inputs as they flow through the layers, and none of them seems to do anything unexpected:
<img width="413" alt="input layer" src="https://user-images.githubusercontent.com/21970939/156879904-a9e3ba08-f5ac-498a-b096-ab60b72ea43d.PNG">
<img width="397" alt="after image_encoder" src="https://user-images.githubusercontent.com/21970939/156879909-bd68234f-2fea-4be6-9acb-9d7915184e76.PNG">
<img width="401" alt="after temp_encoder" src="https://user-images.githubusercontent.com/21970939/156880294-6c2d8eed-8522-4679-b90e-8ebc9046f5e1.PNG">
<img width="450" alt="after agg" src="https://user-images.githubusercontent.com/21970939/156879913-9a0b3293-2994-4984-9c62-2ad2e2afa396.PNG">
<img width="862" alt="after head" src="https://user-images.githubusercontent.com/21970939/156879920-6b2dd19f-1a29-4f7d-93c2-36471a6d8e9b.PNG">
<img width="853" alt="after softmax" src="https://user-images.githubusercontent.com/21970939/156879922-d9c321be-438e-405b-8ccc-06865233af76.PNG">
<img width="901" alt="output vs ground truth" src="https://user-images.githubusercontent.com/21970939/156880155-fd4013c0-cedc-4215-9220-924a9d7783ee.PNG">

I'm out of ideas and would appreciate any input.

**To Reproduce**
1. Clone https://github.com/ValterFallenius/metnet
2. Download data samples from README link
3. Install requirements
4. Run the code

**Comment by JackKelly** [comment]
Great work plotting all this information!

I'm afraid I'm not familiar enough with the code or the model to really help. But, in the meantime, would it be at all possible to share a link to Weights and Biases for this training run, please? To help us debug.

**Comment by bmkor** [comment]
Sorry for my chip in.  @ValterFallenius 
Wondering where did you get the sample data? both [openclimatefix/goes](https://huggingface.co/datasets/openclimatefix/goes) and [openclimatefix/mrms](https://huggingface.co/datasets/openclimatefix/mrms) in huggingface or somewhere else? Mind sharing me a bit? Wanna give a try as well. Thanks in advance.

**Comment by ValterFallenius** [comment]
Hey @bmkor,

I am using neither actually, you can find my raw data in #27. However I have not published elevation, longitude/latitude data I have used, let me know if you need it. But unless you are writing your thesis for the swedish government I think you might be better off using the original dataset available on huggingface ^^

/Valter

**Comment by bmkor** [comment]
Thanks a lot for your prompt reply and comment. Would try to use those available in the huggingface first. See if I can make the model run.

**Comment by jacobbieker** [comment]
> Thanks a lot for your prompt reply and comment. Would try to use those available in the huggingface first. See if I can make the model run.

Hi, just so you know, the goes dataset currently doesn't have data in it, I'm working through adding data for that. The MRMS dataset does, although I am still finishing up the dataset script. But if you want to get started with that radar data, you can just download the Zarr files themselves and open them locally quite easily.

**Comment by CUITCHENSIYU** [comment]
hello, can you tell me where can download MRMS dataset? thank you vary much!
